### PR TITLE
SDP basic tests support

### PR DESCRIPTION
**Pull request by etienne-lms**
Dependency on :
- ~https://github.com/OP-TEE/optee_test/pull/181~ (merged)
- ~https://github.com/OP-TEE/optee_os/pull/1410~ (merged)
- ~https://github.com/OP-TEE/optee_os/pull/1412 (SDP basics and support in TAs)~  (merged)
- ~https://github.com/OP-TEE/optee_os/pull/1413 (SDP for pTAs)~ (merged)
- ~https://github.com/OP-TEE/optee_os/pull/1414 (cache maintenance services)~  (merged)

### CODE REVIEW

File: Android.mk [hunk 1]
```diff
@@ -69,6 +69,13 @@ LOCAL_C_INCLUDES += $(LOCAL_PATH)/host/xtest \
 		$(LOCAL_PATH)/ta/aes_perf/include \
 		$(LOCAL_PATH)/ta/socket/include
 
+ifeq ($(CFG_SECURE_DATA_PATH),y)
+CFLAGS += -DWITH_SECURE_DATA_PATH
```
Comment (jforissier) [inline Android.mk]: Use `CFG_SECURE_DATA_PATH` instead of `WITH_SECURE_DATA_PATH`, then you don't need to set `CFLAGS`.

File: Android.mk [hunk 2]
```diff
@@ -69,6 +69,13 @@ LOCAL_C_INCLUDES += $(LOCAL_PATH)/host/xtest \
...
+
+srcs += sdp_basic.c
+LOCAL_C_INCLUDES += $(LOCAL_PATH)/ta/sdp_basic/include
```
Comment (jforissier) [inline Android.mk]: Add it always to `LOCAL_C_INCLUDES` for simplicity (same as `socket` which is also an optional feature of OP-TEE OS)

File: host/xtest/Makefile [hunk 3]
```diff
@@ -56,6 +56,12 @@ srcs += xtest_4000.c \
 	adbg/src/adbg_run.c \
 	adbg/src/security_utils_hex.c
 
+ifeq ($(CFG_SECURE_DATA_PATH),y)
+CFLAGS += -DWITH_SECURE_DATA_PATH
```
Comment (jforissier) [inline host/xtest/Makefile]: Same as above

File: host/xtest/Makefile [hunk 4]
```diff
@@ -87,6 +93,9 @@ CFLAGS += -I../../ta/concurrent_large/include
 CFLAGS += -I../../ta/sha_perf/include
 CFLAGS += -I../../ta/aes_perf/include
 CFLAGS += -I../../ta/socket/include
+ifeq ($(CFG_SECURE_DATA_PATH),y)
```
Comment (jforissier) [inline host/xtest/Makefile]: No need to make this conditional

File: ta/sdp_basic/Android.mk [hunk 5]
```diff
@@ -0,0 +1,8 @@
+LOCAL_PATH := $(call my-dir)
+
+ifeq ($(CFG_CACHE_API),y)
+CFLAGS += -DCFG_CACHE_API=y
+endif
```
Comment (jforissier) [inline ta/sdp_basic/Android.mk]: Not needed, `CFG_CACHE_API` should come from `conf.mk` exported by the optee_os build.

File: ta/sdp_basic/Makefile [hunk 6]
```diff
@@ -0,0 +1,8 @@
+BINARY = 12345678-5b69-11e4-9dbb101f74f00099
+
+ifeq ($(CFG_CACHE_API),y)
+CFLAGS += -DCFG_CACHE_API=y
+endif
```
Comment (jforissier) [inline ta/sdp_basic/Makefile]: Same above.

File: host/xtest/sdp_basic.c [hunk 7]
```diff
@@ -0,0 +1,521 @@
...
+ * into the buffer, the CA uses one of the random bytes to set the value of the
+ * offset in the accessed secure buffer.
+ *
+ * For debugging support, the CA may map (in nonsecure world) the secure
+ * buffer to read its content. As this is unsafe on a hardened platform, this
+ * operation is default disable. When enable, error only print out a warning
+ * trace but does not actually fail the test. This also give an easy way to
+ * check that some HW complains on access violation when nonsecure accesses
+ * secure data.
+ */
```
Comment (jforissier) [inline host/xtest/sdp_basic.c]: Nice documentation :+1:

File: host/xtest/sdp_basic.c [hunk 8]
```diff
@@ -0,0 +1,521 @@
...
+#define BUF_MUST_MAP	0
+
+#define DEFAULT_ION_HEAP_TYPE	ION_HEAP_TYPE_UNMAPPED
+
+/*
+ * Warm when nonsecure maps SDP buffer and found unexpected content.
+ * Since mapping SDP buffers from nonsecure is unsafe, such accesses
+ * from nonsecure shall not trigger errors, only warnings.
+ */
+static int warm_ns_access = 1;
```
Comment (jforissier) [inline host/xtest/sdp_basic.c]: s/Warm/Warn/
a/warm/warn/

Is this variable actually used somewhere? I can't find it.

File: ta/sdp_basic/Android.mk [hunk 9]
```diff
@@ -0,0 +1,8 @@
...
+
+local_module := 12345678-5b69-11e4-9dbb101f74f00099.ta
```
Comment (jforissier) [inline ta/sdp_basic/Android.mk]: Wrong format for UUID (missing `-`), should be: `12345678-5b69-11e4-9dbb-101f74f00099.ta`

File: ta/sdp_basic/Makefile [hunk 10]
```diff
@@ -0,0 +1,8 @@
+BINARY = 12345678-5b69-11e4-9dbb101f74f00099
```
Comment (jforissier) [inline ta/sdp_basic/Makefile]: Same as above

File: Android.mk (hunk 1 above)
Comment (etienne-lms) [inline Android.mk]: You mean change code inside TAs implementation to rely straight on the `CFG_SECURE_DATA_PATH` from the devkit config ? I will do that.

File: Android.mk (hunk 2 above)
Comment (etienne-lms) [inline Android.mk]: ok

File: host/xtest/Makefile (hunk 3 above)
Comment (etienne-lms) [inline host/xtest/Makefile]: CFG_xxx are visible from host test source files ?

File: Android.mk (hunk 1 above)
Comment (jforissier) [inline Android.mk]: Yes, exactly.

File: host/xtest/Makefile (hunk 3 above)
Comment (jforissier) [inline host/xtest/Makefile]: Yes, thanks to https://github.com/OP-TEE/optee_test/blob/master/host/xtest/Makefile#L11.

File: host/xtest/sdp_basic.c (hunk 8 above)
Comment (etienne-lms) [inline host/xtest/sdp_basic.c]: yep, actually i removed the related code. I will remove this.

File: ta/sdp_basic/Android.mk (hunk 5 above)
Comment (etienne-lms) [inline ta/sdp_basic/Android.mk]: ok

File: ta/sdp_basic/Android.mk (hunk 9 above)
Comment (etienne-lms) [inline ta/sdp_basic/Android.mk]: thanks

File: ta/sdp_basic/Makefile (hunk 10 above)
Comment (etienne-lms) [inline ta/sdp_basic/Makefile]: thanks

File: ta/sdp_basic/Makefile (hunk 6 above)
Comment (etienne-lms) [inline ta/sdp_basic/Makefile]: ok

File: ta/sdp_perf/Android.mk [hunk 11]
```diff
@@ -0,0 +1,8 @@
+LOCAL_PATH := $(call my-dir)
+
+ifeq ($(CFG_CACHE_API),y)
+CFLAGS += -DCFG_CACHE_API=y
+endif
```
Comment (etienne-lms) [inline ta/sdp_perf/Android.mk]: wil remove this

File: ta/sdp_perf/Android.mk [hunk 12]
```diff
@@ -0,0 +1,8 @@
...
+
+local_module := 12345678-5b69-11e4-9dbb101f74f00066.ta
```
Comment (etienne-lms) [inline ta/sdp_perf/Android.mk]: will add `-` inside UUID

File: ta/sdp_perf/Makefile [hunk 13]
```diff
@@ -0,0 +1,8 @@
+BINARY = 12345678-5b69-11e4-9dbb101f74f00066
```
Comment (etienne-lms) [inline ta/sdp_perf/Makefile]: will add `-` inside UUID

File: ta/sdp_perf/Makefile [hunk 14]
```diff
@@ -0,0 +1,8 @@
...
+
+ifeq ($(CFG_CACHE_API),y)
+CFLAGS += -DCFG_CACHE_API=y
+endif
```
Comment (etienne-lms) [inline ta/sdp_perf/Makefile]: will remove this

**Comment by etienne-lms** [comment]
Comments addressed.
SDP perf test removed from this PR.
SDP pTA test included in this PR (wrongly submitted, yet nicely reviewed in https://github.com/OP-TEE/optee_test/pull/183).

**Comment by jforissier** [comment]
Looks good to me!
`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

**Comment by etienne-lms** [comment]
rebased, no change.

**Comment by jforissier** [comment]
Since you're exporting the pTA commands in `pta_invoke_tests.h` (optee_os commit 620e4edf164b), don't you think you should use them instead of re-defining the commands locally?

**Comment by etienne-lms** [comment]
sure. i have a patch ready for that. I can push it on top of this PR...

**Comment by jforissier** [comment]
Yes, please, in fact if you can squash it into "xtest: SDP basic tests TA to pseudo TA with SDP memref" it would be even better.

**Comment by jforissier** [comment]
All good.
- For "xtest/sdp: test NS client invocation to pTA with SDP memref params": `Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`
- For "xtest: 'invoke tests' pTA API from exported include files": `Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>` and as I said I'd rather have it squashed.

**Comment by etienne-lms** [comment]
Oups, sorry, i pushed back a squashed version as you required while you were reviewing.
Yet, changes are the same and I understood you agree with them. So I will now apply the tags.
Thanks for the review.

**Comment by jforissier** [comment]
611ea2e should be squashed into 101540e?

**Comment by etienne-lms** [comment]
I would rather keep them separated. Later commit is now only dedicated to xtest 1001 while the former patch specifically targets SDP tests against a pTA.

**Comment by etienne-lms** [comment]
tags applied. (feel free to keep on commenting if required)

**Comment by jforissier** [comment]
@etienne-lms ah yes sorry it's xtest 1001, so it's fine. Would have been good to mention "xtest 1001" in the commit log ;) but it's OK.

**Comment by etienne-lms** [comment]
i can still change the commit comment...

**Pull request merged**